hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/metadata/HazelcastRelMdPrunability.java

```diff
@@ -0,0 +1,177 @@
+import com.google.common.collect.ImmutableSet;
+import com.hazelcast.jet.sql.impl.opt.OptUtils;
+import com.hazelcast.jet.sql.impl.opt.metadata.HazelcastRelMdPrunability.PrunabilityMetadata;
+import com.hazelcast.jet.sql.impl.opt.physical.FullScanPhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.physical.IndexScanMapPhysicalRel;
+import com.hazelcast.jet.sql.impl.opt.prunability.PartitionStrategyConditionExtractor;
+import com.hazelcast.jet.sql.impl.schema.HazelcastTable;
+import com.hazelcast.sql.impl.extract.QueryPath;
+import com.hazelcast.sql.impl.schema.TableField;
+import com.hazelcast.sql.impl.schema.map.PartitionedMapTable;
+import org.apache.calcite.linq4j.tree.Types;
+import org.apache.calcite.plan.volcano.RelSubset;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.Aggregate;
+import org.apache.calcite.rel.core.Calc;
+import org.apache.calcite.rel.core.Join;
+import org.apache.calcite.rel.metadata.Metadata;
+import org.apache.calcite.rel.metadata.MetadataDef;
+import org.apache.calcite.rel.metadata.MetadataHandler;
+import org.apache.calcite.rel.metadata.ReflectiveRelMetadataProvider;
+import org.apache.calcite.rel.metadata.RelMetadataProvider;
+import org.apache.calcite.rel.metadata.RelMetadataQuery;
+import org.apache.calcite.rex.RexCall;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.util.Util;
+
+import java.lang.reflect.Method;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.Collections.emptyMap;
+
+public final class HazelcastRelMdPrunability
+        implements MetadataHandler<PrunabilityMetadata> {
+
+    public static final RelMetadataProvider SOURCE = ReflectiveRelMetadataProvider.reflectiveSource(
+            PrunabilityMetadata.METHOD,
+            new HazelcastRelMdPrunability()
+    );
+
+    private HazelcastRelMdPrunability() {
+    }
+
+    @Override
+    public MetadataDef<PrunabilityMetadata> getDef() {
+        return PrunabilityMetadata.DEF;
+    }
+
+    @SuppressWarnings("unused")
+    public Map<String, List<Map<String, RexNode>>> extractPrunability(
+            FullScanPhysicalRel scan,
+            RelMetadataQuery mq
+    ) {
+
+        final HazelcastTable hazelcastTable = OptUtils.extractHazelcastTable(scan);
+        if (!(hazelcastTable.getTarget() instanceof PartitionedMapTable)) {
+            return emptyMap();
+        }
+
+        final PartitionedMapTable targetTable = hazelcastTable.getTarget();
+        Set<String> partitioningColumns;
+
+        if (targetTable.partitioningAttributes().isEmpty()) {
+            partitioningColumns = ImmutableSet.of(QueryPath.KEY);
+        } else {
+            final HashSet<String> partitioningFieldNames = new HashSet<>(targetTable.partitioningAttributes());
+            partitioningColumns = targetTable.keyFields()
```

Comment:
what does this filtering do? it seems to me that it only checks consistency between `keyFields` and `partitioningColumns`. Can there ever be a discrepancy. BTW, I reordered the expression to make it more explicit what it does (and better for common case without partitioning attributes)

Reply:
PartitioningColumns contains field names rather than columns names, we have to convert it to column names here. There's a "test" for this in PartitionPruningIT actually (`test_renamingKeyPruned`).

Reply:
aha, so this implements `EXTERNAL NAME`?

Reply:
I added a comment.